SU2_CFD/src/solvers/CGradientSmoothingSolver.cpp

```diff
@@ -0,0 +1,1168 @@
+      my_Gradient = SU2_TYPE::GetDerivative(DV_Value[iDV][iDV_Value]);
+
+      SU2_MPI::Allreduce(&my_Gradient, &localGradient, 1, MPI_DOUBLE, MPI_SUM, SU2_MPI::GetComm());
+
+      output[iDV_index] = localGradient;
+      iDV_index++;
+    }
+  }
+
+  AD::ClearAdjoints();
+
+}
+
+void CGradientSmoothingSolver::SetSensitivity(CGeometry *geometry, CConfig *config, CSolver*) {
+  unsigned long iPoint;
+  unsigned int iDim;
+  for (iPoint = 0; iPoint < nPoint; iPoint++) {
+    for (iDim = 0; iDim < nDim; iDim++) {
+      // nodes->SetSensitivity(iPoint,iDim, solver->GetNodes()->GetSensitivity(iPoint,iDim));
+    }
+  }
+}
+
+void CGradientSmoothingSolver::WriteSensitivity(CGeometry *geometry, CConfig *config, unsigned long val_marker) {
+
+  unsigned long iPoint, total_index;
+  unsigned int iDim;
+  su2double* normal;
+  su2double norm;
+
+  /*--- split between surface and volume first, to avoid mpi ranks with no part of the marker to write back nonphysical solutions in the surface case ---*/
+  if ( config->GetSmoothOnSurface() ) {
+    if( val_marker!=NOT_AVAILABLE ) {
+      for (unsigned long iVertex =0; iVertex<geometry->nVertex[val_marker]; iVertex++)  {
+        iPoint = geometry->vertex[val_marker][iVertex]->GetNode();
+        normal = geometry->vertex[val_marker][iVertex]->GetNormal();
+        norm = 0.0;
+        for (iDim = 0; iDim < nDim; iDim++) {
+          norm += normal[iDim]*normal[iDim];
+        }
+        norm = sqrt(norm);
+        for (iDim = 0; iDim < nDim; iDim++) {
+          normal[iDim] = normal[iDim] / norm;
+        }
+        for (iDim = 0; iDim < nDim; iDim++) {
+          this->GetNodes()->SetSensitivity(iPoint, iDim, normal[iDim]*LinSysSol[iPoint]);
+        }
+      }
+    }
+  } else {
+    if ( config->GetSepDim() ) {
+      for (iPoint = 0; iPoint < nPoint; iPoint++) {
+        this->GetNodes()->SetSensitivity(iPoint, dir, LinSysSol[iPoint]);
+      }
+    } else {
+      for (iPoint = 0; iPoint < nPoint; iPoint++) {
+        for (iDim = 0; iDim < nDim; iDim++) {
+          total_index = iPoint*nDim + iDim;
+          this->GetNodes()->SetSensitivity(iPoint, iDim, LinSysSol[total_index]);
+        }
+      }
+    }
+  }
+}
+
+void CGradientSmoothingSolver::WriteSens2Geometry(CGeometry *geometry, CConfig *config) {
+  unsigned long iPoint;
+  unsigned int iDim;
+  for (iPoint = 0; iPoint < nPoint; iPoint++) {
+    for (iDim = 0; iDim < nDim; iDim++) {
+      geometry->SetSensitivity(iPoint,iDim, nodes->GetSensitivity(iPoint,iDim));
+    }
+  }
+}
+
+void CGradientSmoothingSolver::ReadSens2Geometry(CGeometry *geometry, CConfig *config) {
+  unsigned long iPoint;
+  unsigned int iDim;
+  for (iPoint = 0; iPoint < nPoint; iPoint++) {
+    for (iDim = 0; iDim < nDim; iDim++) {
+      nodes->SetSensitivity(iPoint, iDim, geometry->GetSensitivity(iPoint,iDim));
+    }
+  }
+}
+
+void CGradientSmoothingSolver::WriteVector2Geometry(CGeometry *geometry, CConfig *config, CSysVector<su2matvecscalar>& vector) {
```

Comment:
Yes, this could be a function of CGeometry, since theoretically nothing from outside is needed here
This would mean including CSysVector in CGeometry and adding 2 new functions which are only called by one solver?
Is this okay, or would it be too much change to a basic class like CGeometry?

Reply:
You don't need to include it, make the function a template to accept anything with operator (i, j).